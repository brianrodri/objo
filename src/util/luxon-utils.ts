import assert from "assert";
import { DateTime, DateTimeOptions, Duration, Interval } from "luxon";
import { Brand } from "utility-types";

/** Union of luxon types that can be checked for validity. */
export type LuxonValue<IsValid extends boolean> = DateTime<IsValid> | Duration<IsValid> | Interval<IsValid>;

/** A format string that can correctly format and parse {@link DateTime}s. */
export type LuxonFormat = Brand<string, "LuxonFormat">;

/**
 * Asserts that the provided Luxon value is valid.
 *
 * This function checks whether a Luxon date, duration, or interval is valid. If the value is invalid,
 * it throws an error with a message combining a custom header (or default constructor name) with the
 * value's invalid reason and, if available, its invalid explanation.
 * @param value - The Luxon object to validate.
 * @param message - Optional custom header for the error message.
 * @throws If the provided value is invalid.
 */
export function assertValidLuxonValue(
    value: LuxonValue<true> | LuxonValue<false>,
    message?: string,
): asserts value is LuxonValue<true> {
    message ??= `Invalid ${value.constructor.name}`;
    const help = value.invalidExplanation ? `${value.invalidReason}: ${value.invalidExplanation}` : value.invalidReason;
    assert(value.isValid, `${message}: ${help}`);
}

/**
 * Asserts that the provided date format yields consistent results when formatting and parsing a date.
 *
 * This function formats a fixed UTC reference date into a string using the given format and optional options,
 * then parses that string back into a date. It verifies that the parsed date is valid and that reformatting it
 * produces the original string. Failing this consistency check triggers an error, indicating that the format is invalid
 * for round-trip date conversions.
 * @param dateFormat - The date format string to validate.
 * @param dateOptions - Optional formatting and parsing options for date-time operations.
 * @see {@link https://moment.github.io/luxon/#/parsing?id=table-of-tokens}
 * @throws {AssertionError} If the date format does not produce matching formatted strings upon parsing.
 */
export function assertValidDateTimeFormat(
    dateFormat: string,
    dateOptions?: DateTimeOptions,
): asserts dateFormat is LuxonFormat {
    const sourceDate = DateTime.fromMillis(0).setZone("utc"); // NOTE: Any valid UTC date works.
    const parsedDate = DateTime.fromFormat(sourceDate.toFormat(dateFormat, dateOptions), dateFormat, dateOptions);
    assert(
        parsedDate.isValid &&
            parsedDate.toFormat(dateFormat, dateOptions) === sourceDate.toFormat(dateFormat, dateOptions),
        "date format must parse the formatted strings it produces",
    );
}
