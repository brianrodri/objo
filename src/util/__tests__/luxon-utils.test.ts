import { DateTime, DateTimeMaybeValid, Duration, DurationMaybeValid, Interval, IntervalMaybeValid } from "luxon";
import { describe, expect, expectTypeOf, it } from "vitest";

import { assertValidDateTimeFormat, assertValidLuxonValue, mergeIntersecting } from "../luxon-utils";
import { MERGE_INTERSECTING_TEST_CASES } from "./luxon-utils.test.const";

describe(`${assertValidLuxonValue.name}`, () => {
    const reason = "user-provided reason";
    const explanation = "user-provided explanation";

    describe.each(["user-provided message", undefined])("with message=%j", (message) => {
        it("should accept valid datetimes", () => {
            const value = DateTime.fromISO("2025-03-05");
            expectTypeOf(value).toEqualTypeOf<DateTimeMaybeValid>();
            assertValidLuxonValue(value, message);
            expectTypeOf(value).toEqualTypeOf<DateTime<true>>();
        });

        it("should accept valid durations", () => {
            const value = Duration.fromDurationLike({ days: 1 }) as DurationMaybeValid;
            expectTypeOf(value).toEqualTypeOf<DurationMaybeValid>();
            assertValidLuxonValue(value, message);
            expectTypeOf(value).toEqualTypeOf<Duration<true>>();
        });

        it("should accept valid intervals", () => {
            const value = Interval.after(DateTime.now(), { days: 1 });
            expectTypeOf(value).toEqualTypeOf<IntervalMaybeValid>();
            assertValidLuxonValue(value, message);
            expectTypeOf(value).toEqualTypeOf<Interval<true>>();
        });

        it("should reject invalid datetimes", () => {
            const value = DateTime.invalid(reason, explanation);
            expect(() => assertValidLuxonValue(value, message)).toThrowErrorMatchingSnapshot();
        });

        it("should reject invalid durations", () => {
            const value = Duration.invalid(reason, explanation);
            expect(() => assertValidLuxonValue(value, message)).toThrowErrorMatchingSnapshot();
        });

        it("should reject invalid intervals", () => {
            const value = Interval.invalid(reason, explanation);
            expect(() => assertValidLuxonValue(value, message)).toThrowErrorMatchingSnapshot();
        });
    });
});

describe(`${assertValidDateTimeFormat.name}`, () => {
    it.each(["yyyy-MM-dd", "kkkk-'W'WW", "yyyy-MM"])("should accept valid format=%j", (validFormat) => {
        expect(() => assertValidDateTimeFormat(validFormat)).not.toThrow();
    });

    it.each(["FF", "''"])("should reject invalid format=%j", (invalidFormat) => {
        expect(() => assertValidDateTimeFormat(invalidFormat)).toThrowErrorMatchingSnapshot();
    });
});

describe(`${mergeIntersecting.name}`, () => {
    it.each(MERGE_INTERSECTING_TEST_CASES)("$name", ({ input, output }) => {
        expect(mergeIntersecting(input)).toEqual(output);
    });
});
