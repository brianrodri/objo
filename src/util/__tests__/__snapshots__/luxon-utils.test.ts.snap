// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assertValid > with message="user-provided message" > should reject invalid datetimes 1`] = `[AssertionError: user-provided message: user-provided reason: user-provided explanation]`;

exports[`assertValid > with message="user-provided message" > should reject invalid durations 1`] = `[AssertionError: user-provided message: user-provided reason: user-provided explanation]`;

exports[`assertValid > with message="user-provided message" > should reject invalid intervals 1`] = `[AssertionError: user-provided message: user-provided reason: user-provided explanation]`;

exports[`assertValid > with message=undefined > should reject invalid datetimes 1`] = `[AssertionError: Invalid DateTime: user-provided reason: user-provided explanation]`;

exports[`assertValid > with message=undefined > should reject invalid durations 1`] = `[AssertionError: Invalid Duration: user-provided reason: user-provided explanation]`;

exports[`assertValid > with message=undefined > should reject invalid intervals 1`] = `[AssertionError: Invalid Interval: user-provided reason: user-provided explanation]`;

exports[`assertValidDateTimeFormat > should reject invalid format="''" 1`] = `[AssertionError: format="''" and options=undefined must make round-trip conversions]`;

exports[`assertValidDateTimeFormat > should reject invalid format="FF" 1`] = `[AssertionError: format="FF" and options=undefined must make round-trip conversions]`;
